autoconf problems for builds with static HDF5 library

This is similar to #324 but in the autotools build.

When shared libraries are used, the HDF5 shared library knows where to find the libraries it needs, so these problems do not occur.

With static only builds, if the HDF5 library is built with szlib, then the netCDF configure will fail like this:


```
checking for library containing H5Fflush... no
configure: error: Can't find or link to the hdf5 library. Use --disable-netcdf-4, or see config.log for errors.

```

A look in config.log reveals this:

```
configure:17782: gcc -o conftest  -I/usr/local/wrf_io_serial_static/include -L/usr/local/wrf_io_serial_static/lib conftest.c -lhdf5  -lm -lz -lcurl  >&5
/usr/local/wrf_io_serial_static/lib/libhdf5.a(H5Z.o): In function `H5Z__init_package':
H5Z.c:(.text+0x75c): undefined reference to `SZ_encoder_enabled'
/usr/local/wrf_io_serial_static/lib/libhdf5.a(H5Zszip.o): In function `H5Z_filter_szip':
H5Zszip.c:(.text+0x126): undefined reference to `SZ_BufftoBuffDecompress'
H5Zszip.c:(.text+0x2df): undefined reference to `SZ_BufftoBuffCompress'
/usr/local/wrf_io_serial_static/lib/libhdf5.a(H5PLint.o): In function `H5PL__close':
H5PLint.c:(.text+0x4f8): undefined reference to `dlclose'
/usr/local/wrf_io_serial_static/lib/libhdf5.a(H5PLint.o): In function `H5PL__open':
H5PLint.c:(.text+0x552): undefined reference to `dlopen'
H5PLint.c:(.text+0x56d): undefined reference to `dlsym'
H5PLint.c:(.text+0x605): undefined reference to `dlerror'
/usr/local/wrf_io_serial_static/lib/libhdf5.a(H5PLplugin_cache.o): In function `H5PL__find_plugin_in_cache':
H5PLplugin_cache.c:(.text+0x32a): undefined reference to `dlsym'
collect2: error: ld returned 1 exit status
configure:17782: $? = 1
```

The problem here is that -lsz does not appear in the compile line, so the static only HDF5 library can't find it.

The configure.ac is supposed to be checking for the szlib when necessary:

  ```
 # The user may have built HDF5 with the SZLIB library.
   enable_szlib=no
   if test "x$ac_cv_func_H5Z_SZIP" = xyes; then
    enable_szlib=yes
    AC_SEARCH_LIBS([ SZ_Compress], [szip sz], [], [])
  	AC_DEFINE([USE_SZIP], [1], [if true, compile in szip compression in netCDF-4 variables])
   fi

```

But this happens too late.

Also, the HDF5 library uses the dynamic loading libraries. (See this post from Elena on the HDF forum: http://hdf-forum.184993.n3.nabble.com/Errors-compiling-against-Static-build-HDF5-1-8-11-Need-for-ldl-added-to-linker-arguments-td4026300.html)

I have a fix for these problems and will put up a PR shortly.

--- a/configure.ac
+++ b/configure.ac
@@ -13,7 +13,7 @@
 AC_PREREQ([2.59])
 
 # Initialize with name, version, and support email address.
-AC_INIT([netCDF], [4.6.3-development], [support-netcdf@unidata.ucar.edu])
+AC_INIT([netCDF], [4.6.3-development], [support-netcdf@unidata.ucar.edu], [netcdf-c])
 
 ##
 # Prefer an empty CFLAGS variable instead of the default -g -O2.
@@ -994,6 +994,8 @@ if test "x$enable_netcdf_4" = xyes || test "x$enable_dap" = xyes; then
    AC_SEARCH_LIBS([deflate], [zlibwapi zlibstat zlib zlib1 z], [], [
      AC_MSG_ERROR([Can't find or link to the z library. Turn off netCDF-4 and \
      DAP clients with --disable-netcdf-4 --disable-dap, or see config.log for errors.])])
+   AC_SEARCH_LIBS([SZ_Compress], [szip sz], [], [])
+   AC_SEARCH_LIBS([dlopen], [dl dld], [], [])
 fi
 
 # We need the math library
@@ -1050,17 +1052,14 @@ if test "x$enable_netcdf_4" = xyes; then
       fi
    fi
 
-
-
    # The user may have built HDF5 with the SZLIB library.
+   AC_MSG_CHECKING([whether szlib was used when building HDF5])
    enable_szlib=no
    if test "x$ac_cv_func_H5Z_SZIP" = xyes; then
-    enable_szlib=yes
-    AC_SEARCH_LIBS([ SZ_Compress], [szip sz], [], [])
-  	AC_DEFINE([USE_SZIP], [1], [if true, compile in szip compression in netCDF-4 variables])
+      enable_szlib=yes
+      AC_DEFINE([USE_SZIP], [1], [if true, compile in szip compression in netCDF-4 variables])
    fi
-
-
+   AC_MSG_RESULT([$enable_szlib])
 
    if test "x$ac_cv_func_H5free_memory" = xyes; then
       AC_DEFINE([HDF5_HAS_H5FREE], [1], [if true, H5free_memory() will be used to free hdf5-allocated memory in nc4file.])
@@ -1095,13 +1094,11 @@ fi
 
 # Should we provide parallel io for netcdf-4?
 if test "x$enable_netcdf_4" = xyes ; then
-   AC_MSG_CHECKING([whether parallel I/O is enabled for netcdf-4])
    AC_ARG_ENABLE([parallel4],
 	[AS_HELP_STRING([--disable-parallel4],
                         [disable parallel I/O for netcdf-4, even if it's enabled in libhdf5])],
 			[user_set_parallel4=${enableval}])
    test "x$enable_parallel4" = xno || enable_parallel4=yes
-   AC_MSG_RESULT($enable_parallel4)
 
    # If user wants parallel IO for netCDF-4, make sure HDF5 can provide it.
    if test "x$enable_parallel4" = xyes; then
@@ -1118,6 +1115,8 @@ if test "x$enable_netcdf_4" = xyes ; then
 else
    enable_parallel4=no
 fi
+AC_MSG_CHECKING([whether parallel I/O is enabled for netcdf-4])
+AC_MSG_RESULT($enable_parallel4)
 
 # We have already tested for parallel io in netcdf4
 # parallel I/O for CDF-1, 2, and 5 files can also be done through PnetCDF

